# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importa pandas
import pandas as pd

<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Lee el archivo y almacénalo en df
df=pd.read_csv('../data/music_project_en.csv')

<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
# Obtén la información general sobre nuestros datos
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0     userID  65079 non-null  str  
 1   Track     63736 non-null  str  
 2   artist    57512 non-null  str  
 3   genre     63881 non-null  str  
 4     City    65079 non-null  str  
 5   time      65079 non-null  str  
 6   Day       65079 non-null  str  
dtypes: str(7)
memory usage: 3.5 MB


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay 65079 entradas, pero tanto 'Track', 'artist', 'genre' tiene menos valores no nulos, se tiene que idintificar los valores unicos, si hay diplicados, y de los valores nulos sustirlos si se puede. 




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Se puede interpretar que los datos que hay en las filas, es el registro de cuando un usuario escucho una canción en especial, con su horario, día y lugar donde la escuchó además de los datos de la cannción, titulo, artista y genero. Para saber que alamacenan las columnas. se puede utilizar los metodos .info(), .head() y .sample()

2. Se necesita trabajar más el dataset para obtener información y comprobar la hipótesis 

3. Hay 65079 entradas pero 'Track', 'artist', 'genre' tiene menos valores no nulos, por lo tanto hay valores ausentes o duplicados. 

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='str')

<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle que itera sobre los encabezados y los pone todos en minúsculas
encabezados_minus = []
for c in df.columns:
    encabezados_minus.append(str(c).lower())
df.columns= encabezados_minus   
df.columns

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='str')

<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [7]:
# Bucle que itera sobre los encabezados y elimina los espacios
encabezados_limpio=[]
for c in df.columns:
    encabezados_limpio.append(str(c).strip())
df.columns= encabezados_limpio   
df.columns    

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambia el nombre de la columna "userid"
df.columns= df.columns.str.replace('userid','user_id')
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprueba el resultado: lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [10]:
# Calcula el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace=['track', 'artist', 'genre']
for c in columns_to_replace:
    df[c]=df[c].fillna('unknown')

<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [12]:
# Cuenta los valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [13]:
# Cuenta los duplicados explícitos
print(df.duplicated().sum())

3826


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Elimina los duplicados explícitos
df=df.drop_duplicates().reset_index(drop=True)

<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprueba de nuevo si hay duplicados
print(df.duplicated().sum())

0


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspecciona los nombres de géneros únicos
generos=df['genre'].sort_values(ascending=True).unique()

print(generos)

df['genre'].nunique(dropna=True)



<StringArray>
[       'acid',    'acoustic',      'action',       'adult',      'africa',
   'afrikaans', 'alternative',     'ambient',   'americana',    'animated',
 ...
       'urban',       'uzbek',     'variété',          'vi',   'videogame',
       'vocal',     'western',       'world',   'worldbeat',         'ïîï']
Length: 269, dtype: str


269

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [17]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_values(df: pd.DataFrame, column: str, wrong_values: list, correct_value):
    if column not in df.columns:
        print(f"La columna '{column}' no existe")

    for w in wrong_values:
        df.loc[:,column]=df[column].replace(w, correct_value)

    return df

<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [18]:
# Elimina los duplicados implícitos
incorrectos=['hip', 'hop', 'hip-hop']
replace_wrong_values(df,'genre',incorrectos,'hiphop')


,user_id,track,artist,genre,city,time,day
0,FFB692EC,Kamigata To Boots,The Mass Missile,rock,Shelbyville,20:28:33,Wednesday
1,55204538,Delayed Because of Accident,Andreas Rönnberg,rock,Springfield,14:07:09,Friday
2,20EC38,Funiculì funiculà,Mario Lanza,pop,Shelbyville,20:58:07,Wednesday
3,A3DD03C9,Dragons in the Sunset,Fire + Ice,folk,Shelbyville,08:37:09,Monday
4,E2DC1FAE,Soul People,Space Echo,dance,Springfield,08:34:34,Monday
...,...,...,...,...,...,...,...
61248,729CBB09,My Name,McLean,rnb,Springfield,13:32:28,Wednesday
61249,D08D4A55,Maybe One Day (feat. Black Spade),Blu & Exile,hiphop,Shelbyville,10:00:00,Monday
61250,C5E3A0D5,Jalopiina,unknown,industrial,Springfield,20:09:26,Friday
61251,321D0506,Freight Train,Chas McDevitt,rock,Springfield,21:43:59,Friday


<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprueba de nuevo los duplicados implícitos
unicos = df['genre'].sort_values(ascending=True).unique()

print(unicos)

<StringArray>
[       'acid',    'acoustic',      'action',       'adult',      'africa',
   'afrikaans', 'alternative',     'ambient',   'americana',    'animated',
 ...
       'urban',       'uzbek',     'variété',          'vi',   'videogame',
       'vocal',     'western',       'world',   'worldbeat',         'ïîï']
Length: 266, dtype: str


<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [20]:
# Cuenta las canciones reproducidas en cada ciudad
dias=['Monday', 'Friday']
df_filtrado = df.query("day in @dias")
df_lunes = df.query("day == 'Monday'")
df_viernes = df.query("day == 'Friday'")
print("Canciones que se escucharon lunes y viernes, por ciurdad: ")
print(df_filtrado.groupby(["city"])['track'].count())
print()
print("Canciones que se escucharon el dia lunes, por ciurdad: ")
print(df_lunes.groupby(["city"])['track'].count())
print()
print("Canciones que se escucharon el dia viernes, por ciurdad: ")
print(df_viernes.groupby(["city"])['track'].count())

Canciones que se escucharon lunes y viernes, por ciurdad: 
city
Shelbyville    11509
Springfield    31685
Name: track, dtype: int64

Canciones que se escucharon el dia lunes, por ciurdad: 
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64

Canciones que se escucharon el dia viernes, por ciurdad: 
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

Se puede observar que en Springfield hay mas canciones escuchadas, casi el doble que en Shelbyville. Puede ser que la cantidad de usuarios que usan la aplicación sean más en Springfield. 

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [21]:
# Calcula las canciones reproducidas en cada uno de los dos días
dias=['Monday', 'Friday']
df_filtrado = df.query("day in @dias")

print("Canciones que se escucharon lunes y viernes: ")
print(df_filtrado.groupby(["day"])['track'].count())


Canciones que se escucharon lunes y viernes: 
day
Friday    21840
Monday    21354
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

El día con más actividad es el día viernes. Todavía no logro ver hasta aquí un patron claro. El resultado por día es el mismo si sumamos lo escuchado en ambas ciudades por día. 

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=

    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()

    # Devuelve el número de valores de la columna 'user_id'
def number_tracks(day, city):
    # 1) Filtrar por día
    df_day = df[df['day'] == day]
    # 2) Filtrar por ciudad
    df_city_day = df_day[df_day['city'] == city]
    # 3) Contar apariciones de 'user_id' (cuenta no-nulos)
    return int(df_city_day['user_id'].count())

<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
mondaySpring=number_tracks('Monday', 'Springfield')
print(mondaySpring)

15740


In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
mondayShelby=number_tracks('Monday', 'Shelbyville')
print(mondayShelby)

5614


In [25]:
# El número de canciones reproducidas en Springfield el viernes
fridaySpring=number_tracks('Friday', 'Springfield')
print(fridaySpring)

15945


In [26]:
# El número de canciones reproducidas en Shelbyville el viernes
fridayShelby=number_tracks('Friday', 'Shelbyville')
print(fridayShelby)

5895


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

In [27]:
print("Número de usuarios por ciudad:")
df.groupby('city')['user_id'].nunique()


Número de usuarios por ciudad:


city
Shelbyville    12423
Springfield    29358
Name: user_id, dtype: int64

In [28]:
#top 5 géneros por ciudad
print(df.groupby('city')['genre'].value_counts().groupby(level=0).head(5))

city         genre     
Shelbyville  pop           2431
             dance         1932
             rock          1879
             electronic    1736
             hiphop         960
Springfield  pop           5892
             dance         4435
             rock          3965
             electronic    3786
             hiphop        2096
Name: count, dtype: int64


In [29]:
# Ciudad + día + género
print(df.groupby(['city','day'])['genre'].value_counts().groupby(level=[0,1]).head(3))

city         day        genre
Shelbyville  Friday     pop       748
                        dance     639
                        rock      593
             Monday     pop       732
                        dance     589
                        rock      577
             Wednesday  pop       951
                        rock      709
                        dance     704
Springfield  Friday     pop      2166
                        dance    1602
                        rock     1497
             Monday     pop      2154
                        dance    1669
                        rock     1452
             Wednesday  pop      1572
                        dance    1164
                        rock     1016
Name: count, dtype: int64


In [30]:
# Usuarios más activos
print(df['user_id'].value_counts().head(10))

# Diversidad de géneros por usuario
print(df.groupby('user_id')['genre'].nunique().sort_values(ascending=False).head(10))


user_id
A8AE9169    71
7D166C63    51
414F229D    50
E8339398    43
D6A03F1E    36
EED05948    36
B5496034    34
B851E1A5    33
3E4BFAA7    31
53B9A51D    30
Name: count, dtype: int64
user_id
3D4399B4    18
B5496034    16
7D166C63    15
956422AD    14
107EBCED    14
76E06E09    13
A1261D26    12
9B5F5CDA    12
C7BE9443    11
3E4BFAA7    11
Name: genre, dtype: int64


In [31]:
# Top géneros por día
print(df.groupby('day')['genre'].value_counts().groupby(level=0).head(3))

# Convertir la columna 'time' a formato de hora
df['time'] = pd.to_datetime(df['time'], format='%H:%M:%S').dt.hour

# Agrupar por franja horaria
def franja(hora):
    if 6 <= hora < 12:
        return 'mañana'
    elif 12 <= hora < 18:
        return 'tarde'
    elif 18 <= hora < 24:
        return 'noche'
    else:
        return 'madrugada'

df['franja'] = df['time'].apply(franja)

# Ver géneros por franja
print(df.groupby('franja')['genre'].value_counts().groupby(level=0).head(3))


day        genre
Friday     pop      2914
           dance    2241
           rock     2090
Monday     pop      2886
           dance    2258
           rock     2029
Wednesday  pop      2523
           dance    1868
           rock     1725
Name: count, dtype: int64
franja  genre
mañana  pop      2716
        dance    2050
        rock     1779
noche   pop      2822
        dance    2056
        rock     2053
tarde   pop      2785
        dance    2261
        rock     2012
Name: count, dtype: int64


# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


Según a los datos obtenidos se puede ver que hay una mayor cantidad de usuarios en la ciudad de Springfield a comparación de la ciudad de Shelbyville. En ambas ciudades de escucha más canciones el día viernes que el día lunes. esto podría ser porque los viernes los usuarios acostumbran escuchar más música para relajarse o distrarse de la semana que acabó.   

Al hacer un analisis del top 5 de generos más populares, comparte los mismos gustos ambas ciudades, siendo el genero mas popular pop, seguido del género dance, rock, electronic y hiphop. 

También se hizo un cruce de datos entre Ciudad, día y generos más escuchados, arrojando que el pop, dance y rock son los generos mas escuchados en ambas ciduades los mismos dias.

Otro analisis interesante es la actividada en el día, clasificando los horarios en franjas horarias para analizar la actividad en la mañana, tarde, noche y madrugada, y se puede ver que en la mañana hay mas actividad que en la noche, sin embargo dodne hay menos actividad es en la tarde y de nuevo incrementa la actividad en la noche. Esto nos puede decir los hábitos de los usuarios, les gusta iniciar el día con música para después trabjar con toda la actitud, y en la jornada laboral se concentran hasta salir y en la noche de nuevo escuchan música de seguro para relajarse de la jornada que tubieron los usuarios.  